### ISSUE TITLE
Submission of a JobProcess to a local computer hangs in the TOSUBMIT state (v0.12.2)

### ISSUE THREAD

**Issue by chrisjsewell** [aiidateam/aiida_core#1956]
As discussed at aiidateam/aiida-plugin-cutter#12, submission of a `JobProcess` (but not `WorkChain`) to a local computer (such as ones created for testing) appear to hang in the `TOSUBMIT` state.

For example:

```python
import os
from aiida.orm import Code
from aiida.work.run import run
import aiida_crystal17.tests as tests
from aiida.orm.data.singlefile import SinglefileData

code = Code.get_from_string('crystal17.basic-mock_runcry17@localhost-test')
calc = code.new_calc()
process = calc.process()

infile = SinglefileData(
    file=os.path.join(tests.TEST_DIR, "input_files",
                      'mgo_sto3g_scf.crystal.d12'))
inputs_dict = {
    "_options": {"resources": {"num_machines": 1, "num_mpiprocs_per_machine": 1},
                 "withmpi": False,
                 "max_wallclock_seconds": 30},
    "input_file": infile,
    "code": code
}

output, pid = run(process, _return_pid=True, **inputs_dict)
```

Hangs on:

```console
>>> verdi calculation list 2480
# Last daemon state_updater check: 2D:05h:50m ago (at 09:49:20 on 2018-09-07)
  PK  Creation    State       Sched. state  Computer        Type
----  ----------  --------  --------------  --------------  ---------------
2480  58m ago     TOSUBMIT                  localhost-test  crystal17.basic

```

This only happens for the stable v0.12.2 (and I assume earlier), not the ``develop`` branch.

 An associated example of this occurring in Travis CI is at: https://travis-ci.org/chrisjsewell/aiida-crystal17/builds/426840054

Any suggestions?

**Comment by sphuber** [comment]
Could you find the relevant part of the daemon log? What about if you use `submit` instead of `run`?

**Comment by chrisjsewell** [comment]
Thanks for the quick reply @sphuber. Ok so its not quite the original issue... after I'd made sure everything had been restarted, I can no longer recreate the issue above (although I have a suspicion they are related). But when I run the same code under the `fixture_manager` context, it does hang:

```python
from aiida.utils.fixtures import fixture_manager
with fixture_manager() as fixture_mgr:
    import os
    from aiida.orm import Code
    from aiida_crystal17.tests import TEST_DIR
    import aiida_crystal17.tests.utils as tests
    from aiida_crystal17.utils import run_get_node
    from aiida.orm.data.singlefile import SinglefileData
    computer = tests.get_computer("local-test2", "/Users/cjs14/GitHub/aiida-crystal17/examples/test_workdir", True)
    code = tests.get_code(entry_point='crystal17.basic', computer=computer)
    infile = SinglefileData(
        file=os.path.join(TEST_DIR, "input_files",
                          'mgo_sto3g_scf.crystal.d12'))
    # set up calculation
    calc = code.new_calc()
    inputs_dict = {
        "input_file": infile,
        "code": code,
        "options": {
            "resources": {
                "num_machines": 1,
                "num_mpiprocs_per_machine": 1
            },
            "withmpi": False,
            "max_wallclock_seconds": 30
        }
    }
    process = calc.process()
    calcnode = run_get_node(process, inputs_dict)
('/var/folders/dm/b2qnkb_n3r72slmpxlfmcjvm00lbnd/T/tmpcaIuRt', '/var/folders/dm/b2qnkb_n3r72slmpxlfmcjvm00lbnd/T/tmpcaIuRt/.aiida')
Executing now a migrate command...
for Django backend
Operations to perform:
  Apply all migrations: contenttypes, db, sites, auth, sessions
Running migrations:
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying db.0001_initial... OK
  Applying db.0002_db_state_change... OK
  Applying db.0003_add_link_type... OK
  Applying db.0004_add_daemon_and_uuid_indices... OK
  Applying db.0005_add_cmtime_indices... OK
  Applying db.0006_delete_dbpath... OK
  Applying db.0007_update_linktypes... OK
  Applying db.0008_code_hidden_to_extra... OK
  Applying sessions.0001_initial... OK
  Applying sites.0001_initial... OK
Database was created successfully
Loading new environment...
Installing default AiiDA user...
Starting user configuration for test@aiida.mail...
Configuring a new user with email 'test@aiida.mail'
>> User AiiDA Plugintest saved. <<
** NOTE: no password set for this user, 
         so he/she will not be able to login
         via the REST API and the Web Interface.
Setup finished.
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/calculations/cry_basic.pyc
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/parsers/cry_basic.pyc
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/calculations/cry_basic.pyc
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/calculations/cry_basic.pyc
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/calculations/cry_basic.pyc
/Users/cjs14/GitHub/aiida-crystal17/aiida_crystal17/calculations/cry_basic.pyc
...
```

I'm not sure how to access the daemon log for the test environment?
When I go to `/var/folders/dm/b2qnkb_n3r72slmpxlfmcjvm00lbnd/T/tmpcaIuRt/.aiida`, there is just an empty `daemon/log` folder.

**Comment by chrisjsewell** [comment]
Yes this is principally for testing, `aiida.utils.fixtures.fixture_manager` creates a new database and profile for the duration of the context (destroying it after). That seems reasonable, that maybe a daemon hasn't been started, which I assume the fixture_manager should do? 
As I mentioned, this works fine in aiida-core=1.0, so maybe there has been a fix for this that hasn't been back ported?

**Comment by sphuber** [comment]
I am not very familiar with the test fixtures, that would be @DropD (who left the group unfortunately) and @ltalirz , but as far as I can tell the daemon won't be started  automatically by the manager. The reason that this works nonetheless for `develop` is because the daemon works completely different there. When you `run` something in `develop`, it will be done in its own `runner` which is a mini daemon in and of its own. However, in `0.12` the global daemon is the only one that can submit, update and retrieve job calculations. See if you can start the daemon within the context before running a job calculation and see if that works.

**Comment by chrisjsewell** [comment]
Thanks, that makes a lot more sense to me now.

I quickly tried:

```python
 from aiida.cmdline.commands.daemon import Daemon
    daemon = Daemon()
    daemon.daemon_start()
```

and got:

```
You are not the daemon user! I will not start the daemon.
(The daemon user is 'aiida@localhost', you are 'test@aiida.mail')
```

But, now I know the issue, I should be able to work something out.

**Comment by sphuber** [comment]
In the old version you should configure the daemon user with `verdi daemon configure`, and set that to your test user

**Comment by chrisjsewell** [comment]
Ok sorted thanks, here's my new test fixture:

```python
import os

from aiida.utils.fixtures import fixture_manager
from aiida_crystal17.aiida_compatability import aiida_version, cmp_version
import pytest


@pytest.fixture(scope='session')
def aiida_profile():
    """setup a test profile for the duration of the tests"""
    with fixture_manager() as fixture_mgr:
        yield fixture_mgr


@pytest.fixture(scope='function')
def new_database(aiida_profile):
    """clear the database after each test"""
    yield aiida_profile
    aiida_profile.reset_db()


@pytest.fixture(scope='function')
def new_database_with_daemon(aiida_profile):
    """When you run something in aiida v1, it will be done in its own runner,
    which is a mini daemon in and of its own.
    However, in 0.12 the global daemon is the only one that can
    submit, update and retrieve job calculations.
    Therefore, we must configure and start it before running JobProcesses
    """
    if aiida_version() < cmp_version('1.0.0a1'):
        from aiida.backends.utils import set_daemon_user
        from aiida.cmdline.commands.daemon import Daemon
        from aiida.common import setup

        set_daemon_user(aiida_profile.email)
        daemon = Daemon()
        daemon.logfile = os.path.join(aiida_profile.config_dir,
                                      setup.LOG_SUBDIR, setup.CELERY_LOG_FILE)
        daemon.pidfile = os.path.join(aiida_profile.config_dir,
                                      setup.LOG_SUBDIR, setup.CELERY_PID_FILE)
        daemon.celerybeat_schedule = os.path.join(aiida_profile.config_dir,
                                                  setup.DAEMON_SUBDIR,
                                                  'celerybeat-schedule')

        if daemon.get_daemon_pid() is None:
            daemon.daemon_start()
        else:
            daemon.daemon_restart()
        yield aiida_profile
        daemon.kill_daemon()
        aiida_profile.reset_db()
    else:
        yield aiida_profile
        aiida_profile.reset_db()

```

**Comment by sphuber** [comment]
Fixed in PR #2432
